Add test coverage: update messages in last WFT are written to history

dandavison opened the pull request:
## What changed?
Add test coverage that update messages in the last WFT result in persisted events

## Why?
There was a bug at one point whereby this was not true.

## How did you test it?
Test in PR

```
go test -v -count 1 go.temporal.io/server/tests -run=TestFunctionalSuite/TestUpdateMessageInLastWFT
```

## Potential risks
None

## Is hotfix candidate?
No

dandavison commented:
This wasn't a valid reproduction -- there is a mechanism whereby the messages can be forced to be processed server-side in the required order (thanks @alexshtin). See efc93d6dac140653337ab0ed8676dbff49dc7351

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,113 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+	taskQueueName := "tq"
+	identity := "worker1"
+	workflowType := &commonpb.WorkflowType{Name: workflowTypeName}
+	taskQueue := &taskqueuepb.TaskQueue{Name: taskQueueName, Kind: enumspb.TASK_QUEUE_KIND_NORMAL}
```
I see you are using `tv` in this test. You don't need this vars, you can get values from `tv`.

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,113 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+	identity := "worker1"
+	workflowType := &commonpb.WorkflowType{Name: workflowTypeName}
+	taskQueue := &taskqueuepb.TaskQueue{Name: taskQueueName, Kind: enumspb.TASK_QUEUE_KIND_NORMAL}
+	updateId := tv.UpdateID("update-id")
```
`update-id` will be already there (it is "type" of the var). Here you should pass a number or don't pass anything if you have just one update in the test.

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,113 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+		WorkflowType:        workflowType,
+		TaskQueue:           taskQueue,
+		Input:               nil,
+		WorkflowRunTimeout:  durationpb.New(100 * time.Second),
+		WorkflowTaskTimeout: durationpb.New(1 * time.Second),
```
For "long" timeouts there is `tv.InfinityTimeout()` - which means "I don't care, I just want it to be long enough". And if you are not depends on WFT timeout, you shouldn't specify it too.

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,113 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+	pollResponse := make(chan error)
+	request := &workflowservice.UpdateWorkflowExecutionRequest{
+		Namespace: s.namespace,
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: workflowId,
+			RunId:      we.RunId,
+		},
```
```suggestion
		WorkflowExecution: tv.WorkflowExecution(),
```

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,113 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+	4 WorkflowTaskCompleted
+	5 WorkflowExecutionUpdateAccepted
+	6 WorkflowExecutionCompleted
+	`, s.getHistory(s.namespace, &commonpb.WorkflowExecution{
+		WorkflowId: workflowId,
+		RunId:      we.RunId,
+	}))
```
```suggestion
	`, s.getHistory(s.namespace, tv.WorkflowExecution(),
	))
```
and save history in the var instead of making another call on the next line.

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,113 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+			updateCount++
+		}
+	}
+	s.Equal(1, updateCount)
```
Why do you need this? If `updateCount` is not `1` history assert will fail.

alexshtin commented on tests/reset_workflow.go:
```diff
@@ -651,6 +652,100 @@ func (s *FunctionalSuite) testResetWorkflowReapplyBuffer(
+		_, err := s.engine.UpdateWorkflowExecution(NewContext(), request)
+		updateResponse <- err
+	}()
+	go func() {
```
I usually use main go routine for this.

alexshtin reviewed (commented):
1. Why did you put it into `reset_workflow.go`?
2. I think there is similar test in `update_workflow.go` already.

dandavison commented:
Thanks @alexshtin! I've moved it to the other file and cleaned up further to use the test utilities better. Let me know if you're aware of a test that its redundant with; I didn't see one.
